### DISCUSSION TITLE
Please allow to build embeddings not only from the files on disk. Accept list of PIL objects as allowed parameter

### DISCUSSION THREAD

**Discussion by frutik** [qdrant/fastembed#264] [Ideas]
Current implementation only allows to build embedding from the images on disk (you build PIL objects from them inside code of the ibrary). Please, allow to pass as a parameter list of pre-creaed PILs

**Comment by joein** [comment]
Hey @frutik 
Unfortunately, it is not possible, since PIL images are not serializable, and it will block parallelization features.
However, it might be possible to allow np arrays

Why do you need to pass PILs ?

**Reply by frutik** [reply]
Well, I not nececcarily need PIL, but I'd like to skip "disk" part in processing images. So, I have a list of urls to process, I get images with http client, put results into PILs, (for example) convert them into greyscale / resize, build embedding, profit...

btw, you convert list into PILs internally. So, why not accept the list already converted (as an option) - https://github.com/qdrant/fastembed/blob/main/fastembed/image/onnx_image_model.py#L56

**Reply by joein** [reply]
When `parallel` setting is set, we read the data from disk and convert it to PIL within each child process, we can't pass PIL to those processes, since it is not serializable.

We would probably need to convert it to array, then convert it back to PIL, but at the moment just forbidding PIL as input seems to be a more proper solution.

Btw, at the moment, custom preprocessing is not supported, but only the default ones (we haven't yet provided a way to disable preprocessing for image models).

We'll try to extend the interface in the future releases.
